nixos/searx: declarative configuration

rnhmjoj opened the pull request:
###### Motivation for this change
The pain of having to manually diff the settings.yml on every update,
like if NixOS were some kind of Arch Linux.

###### Things done

- [x] Tested environment substitution
- [x] Tested configuration merging
- [x] Tested service runs
- [x] Remove searx uid,gids
- [x] Document changes in the release notes
- [x] Run more tests for uWSGI
- [x] Figure out why immediate-uid is necessary
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

rnhmjoj committed:
nixos/searx: declarative configuration

rnhmjoj committed:
nixos/ids: remove reserved searx ids

SuperSandro2000 commented on nixos/modules/services/networking/searx.nix:
```diff
@@ -3,32 +3,110 @@
+    EOF
+
+    # merge the two files
+    ${pkgs.yq-go}/bin/yq m -P -i settings.yml "$default"
```
I don't remember the search config fully but is it completely mergeable? Does that do deep merge? What happens if a list has 5 entries and I want t overwrite them?

rnhmjoj replied:
I asked that myself, but surprisingly it's working as expected. In particular, it's merging the `engine` list like a set.
For example, I'm redefining the API keys for a bunch of engines and the result is applying the changes but keeping every other property.

SuperSandro2000 replied:
I still don't have a good feeling about this but the current way is probably more terrible.

rnhmjoj replied:
Yeah, a bit more testing wouldn't hurt.

rnhmjoj replied:
On a more close inspection I found the bitter truth: it did merge the list of engines but overriding unrelated elements. It looks like it's sorting in alphabetical order by `name` and then merging the respective elements.

rnhmjoj committed:
nixos/tests: add searx test

rnhmjoj commented:
I managed to get the right merging behavior by making `engines` an attrset and casting some `jq` spells.
Also added a test and documentation.

rnhmjoj committed:
searx: make package friendly towards nginx/uWSGI

- Link the /static directory to make it easier to serve
- Export searx module for use in uWSGI (just import the module)

rnhmjoj committed:
nixos/searx: add support for running in uWSGI

rnhmjoj committed:
nixos/doc: add searx changes to the relase notes

rnhmjoj commented:
I went down a rabbit trying to run searx in uWSGI but I finally did it.
This is a must because searx runs in werkzeug by default: it logs every request (including search keywords!) and there is no way to disable logging unless changes the webserver, of course.

rnhmjoj commented:
The PR got so big, maybe uWSGI changes should be split out.

rnhmjoj commented:
Fixed a couple more issues with uWSGI.

rnhmjoj commented:
For easier review, I've opened a PR with only the uwsgi changes: #105608

rnhmjoj commented:
Rebased. Now that the uWSGI changes have been accepted this builds in master.

rnhmjoj committed:
searx: add NixOS test to passthu.tests
